19/01/15
====

- Continue with the testing of the new method. For now, I am going to use only the method to order the contigs based on the homozygous SNP distribution. 
- Create new genomes by re-writing the model genome script. 
- In order to create replicates for the experiment to be reliable, I am going to create 5 identical genomes for each condition. I am going to test the effect of different sizes from **1-15 Mb and different contig sizes**. The folders without letter contain genomes with 700 contigs and the folders with an A contain genomes with 1300 contigs. For the SNP distribution, I have chosen a high density, **1 SNP per 500 bp** (same number for hm and ht). 
- Once I have the genomes in the shared folder, I run the SDM to create the ordered fasta file. After that, I am adapting the **find_causal_mutation.rb** script in order to create a common **CSV file** that contains all the different candidate SNP positions for all the genomes and the percentage of shift (how far is the candidate SNP from the causal mutation).