Results 28/01
====

1-15 Mb genomes
----
- 5 identical genomes for each condition (replicates)
- Test the effect of different sizes from **1 to 15 Mb and different contig sizes**.  
- For the SNP distribution, I have chosen a high density, **1 SNP per 500 bp** (same number for hm and ht). 

![Image](https://github.com/pilarcormo/small_genomes_SNPs/blob/master/Results/bar_chart.png?raw=true)

####700 contigs
![Image](https://github.com/pilarcormo/small_genomes_SNPs/blob/master/Results/Rplots-1Mb-1000-700contigs.png?raw=true)

####1300 contigs
![Image](https://github.com/pilarcormo/small_genomes_SNPs/blob/master/Results/Rplots-1Mb-1000.1300_contigs.png?raw=true)


30 Mb genomes
----

<table>
  <tr><th>Number of contigs <th>%</th>
  
  <tr><th>1015.2 </th> <th> 0.525776685</th>
  <tr><th> 2028.8 </th> <th> 0.426776667</th>
  <tr><th> 4055</th> <th>0.505681983</th>
</table>

![Image](https://github.com/pilarcormo/small_genomes_SNPs/blob/master/Results/Rplots-30Mb-1500%20contigs-centre.png?raw=true)

####What happens if the mutation is not located in the centre of the model genome? 

<table>

<table>
 <tr><th>Genome length (Mb) <th>%</th> 
 
 <tr><th> 1 <th> 16.35</th> 
 <tr><th> 3 <th> 16.55 </th> 
 <tr><th> 5 <th> 16.60 </th> 
 <tr><th> 7 <th> 16.77 </th> 
 <tr><th> 9 <th> 16.67 </th> 
 <tr><th> 11 <th> 16.70</th> 
<tr><th> 13 <th>16.62  </th> 
<tr><th> 15 <th> 16.98 </th>
<tr><th> 30 <th> 15.81 </th>
    </table>
![Image](https://github.com/pilarcormo/small_genomes_SNPs/blob/master/Results/Rplots-30Mb-left.png?raw=true)
![Image](https://github.com/pilarcormo/small_genomes_SNPs/blob/master/Results/Rplots-30Mb-right.png?raw=true)

